sdk/python/feast/go_server.py

```diff
@@ -0,0 +1,243 @@
+# Copyright 2022 The Feast Authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import atexit
+import os
+import platform
+import socket
+import subprocess
+import time
+from typing import Any, Dict, List, Union
+
+import grpc
+
+import feast
+from feast.errors import (
+    FeatureNameCollisionError,
+    GoSubprocessConnectionFailed,
+    InvalidFeaturesParameterType,
+)
+from feast.feature_service import FeatureService
+from feast.online_response import OnlineResponse
+from feast.protos.feast.serving.ServingService_pb2 import (
+    GetFeastServingInfoRequest,
+    GetOnlineFeaturesRequest,
+)
+from feast.protos.feast.serving.ServingService_pb2_grpc import ServingServiceStub
+from feast.repo_config import RepoConfig
+from feast.type_map import python_values_to_proto_values
+
+
+class GoServer:
```

Comment:
Played around with running the subprocess in a background thread. Below worked for me. It automatically restarts the process and also allows the main process to kill the subprocess when it exits.

server.py
```python
from time import sleep


import random
import string


id = ''.join(random.choice(string.ascii_lowercase) for i in range(10))

count = 0
while True:
    try:
        print(f'Currently running server subprocess with name:{id} for count {count}')
        sleep(1)
        count += 1

        # # Kill the process every 5 iterations
        if count == 5:
            print('Server subprocess kills itself')
            exit(5)
    except KeyboardInterrupt as e:
        print("Cleaning up subprocess")
        exit(0)
```

main.py
```python
import atexit
import subprocess
import threading
import time
import signal


def background_process(stop_server):
    while True:
        # Start process
        proc = subprocess.Popen(['python3', 'server.py'])
        print()
        print(f"Started new server subprocess {proc.pid}")

        # Check to see if process is still running
        while proc.poll() is None and not stop_server():

            # TODO: Use a gRPC health check here to see if the server is up
            # Don't use readline since it will block!
            print(f"doing health check for process {proc.pid}")

            time.sleep(1)

        if stop_server():
            try:
                proc.wait(timeout=2)
            except subprocess.TimeoutExpired:
                pass
            finally:
                proc.kill()
                print('Cleaning up parent thread')
                break

def main():
    stop_server = False

    def stop_server_gracefully_atexit():
        nonlocal stop_server
        stop_server = True

    def stop_server_gracefully_on_signal(signal, frame):
        stop_server_gracefully_atexit()

    atexit.register(stop_server_gracefully_atexit)
    signal.signal(signal.SIGTERM, stop_server_gracefully_on_signal)
    signal.signal(signal.SIGINT, stop_server_gracefully_on_signal)

    t = threading.Thread(target=background_process, args=(lambda: stop_server,))

    t.start()
    try:
        t.join()
    except KeyboardInterrupt as e:
        pass

    print('Cleaning up main process')

main()
```

Reply:
I'm fixing go server implementation by improving on this!

Reply:
Nice!

Reply:
I have implemented a background thread (as a blocking thread instead of polling) but this took a very long time for pytest and the test didn't work properly so I had 2 versions. I will experiment more on this...

Reply:
> as a blocking thread instead of polling

Blocking is ok, as long as it's still possible to send a shutdown command from the main process. The reason I didn't use something like `stdout.readline` was because it wouldn't receive a command from the main process to shut down.

Reply:
going to resolve this since Achal as successfully added `GoServerMonitorThread` and the lifecycle test is functioning properly